src/merkle/smt/full/concurrent/tests.rs

```diff
@@ -465,3 +466,134 @@ fn test_compute_mutations_parallel() {
     assert_eq!(mutations.node_mutations(), control.node_mutations());
     assert_eq!(mutations.new_pairs(), control.new_pairs());
 }
+
+#[test]
+fn test_smt_construction_with_entries_unsorted() {
+    let entries = [
+        (RpoDigest::new([ONE, ONE, Felt::new(2_u64), ONE]), [ONE; 4]),
+        (RpoDigest::new([ONE; 4]), [ONE; 4]),
+    ];
+    let control = Smt::with_entries_sequential(entries).unwrap();
+    let smt = Smt::with_entries(entries).unwrap();
+    assert_eq!(smt.root(), control.root());
+    assert_eq!(smt, control);
+}
+
+fn arb_felt() -> impl Strategy<Value = Felt> {
+    prop_oneof![any::<u64>().prop_map(Felt::new), Just(ZERO), Just(ONE),]
+}
+
+/// Generate entries that are guaranteed to be in different subtrees
+fn generate_cross_subtree_entries() -> impl Strategy<Value = Vec<(RpoDigest, Word)>> {
+    let subtree_offsets = prop::collection::vec(0..(COLS_PER_SUBTREE * 4), 1..100);
+
+    subtree_offsets.prop_map(|offsets| {
+        offsets
+            .into_iter()
+            .map(|base_col| {
+                let key = RpoDigest::new([ONE, ONE, ONE, Felt::new(base_col)]);
+                let value = [ONE, ONE, ONE, Felt::new(base_col)];
+                (key, value)
+            })
+            .collect()
+    })
+}
+
+fn arb_entries() -> impl Strategy<Value = Vec<(RpoDigest, Word)>> {
+    // Combine random entries with guaranteed cross-subtree entries
+    prop_oneof![
+        // Original random entry generation
+        prop::collection::vec(
+            prop_oneof![
+                // Random values case
+                (
+                    prop::array::uniform4(arb_felt()).prop_map(RpoDigest::new),
+                    prop::array::uniform4(arb_felt())
+                ),
+                // Edge case values
+                (
+                    Just(RpoDigest::new([ONE, ONE, ONE, Felt::new(0)])),
+                    Just([ONE, ONE, ONE, Felt::new(u64::MAX)])
+                )
+            ],
+            1..1000,
+        ),
+        // Guaranteed cross-subtree entries
+        generate_cross_subtree_entries(),
+        // Mix of both (combine random and cross-subtree entries)
+        (
+            generate_cross_subtree_entries(),
+            prop::collection::vec(
+                (
+                    prop::array::uniform4(arb_felt()).prop_map(RpoDigest::new),
+                    prop::array::uniform4(arb_felt())
+                ),
+                1..1000,
+            )
+        )
+            .prop_map(|(mut cross_subtree, mut random)| {
+                cross_subtree.append(&mut random);
+                cross_subtree
+            })
+    ]
+    .prop_map(|entries| {
+        // Filter for uniqueness
```

Comment:
Nit: Maybe we can leave a quick note mentioning why we do this (I assume because that's an error condition of `Smt::with_entries`).